vector/src/main/java/im/vector/app/features/matrixto/MatrixToBottomSheetViewModel.kt

```diff
@@ -0,0 +1,192 @@
+                            startChattingState = Success(Unit)
+                    )
+                }
+            }
+            else                      -> {
+                setState {
+                    copy(
+                            matrixItem = Fail(IllegalArgumentException(stringProvider.getString(R.string.unexpected_error))),
+                            startChattingState = Uninitialized
+                    )
+                }
+            }
+        }
+    }
+
+    private suspend fun resolveUser(userId: String): User {
+        return (session.getUser(userId)
+                ?: tryOrNull {
+                    awaitCallback<JsonDict> {
+                        session.getProfile(userId, it)
+                    }
+                }?.let {
+                    User(userId, it[ProfileService.DISPLAY_NAME_KEY] as? String, it[ProfileService.AVATAR_URL_KEY] as? String)
+                }
+                // Create raw Uxid in case the user is not searchable
+                ?: User(userId, null, null))
+    }
+
+    companion object : MvRxViewModelFactory<MatrixToBottomSheetViewModel, MatrixToBottomSheetState> {
+        override fun create(viewModelContext: ViewModelContext, state: MatrixToBottomSheetState): MatrixToBottomSheetViewModel? {
+            val fragment: MatrixToBottomSheet = (viewModelContext as FragmentViewModelContext).fragment()
+            val args: MatrixToBottomSheet.MatrixToArgs = viewModelContext.args()
+
+            return fragment.matrixToBottomSheetViewModelFactory.create(state, args)
+        }
+    }
+
+    override fun handle(action: MatrixToAction) {
+        when (action) {
+            is MatrixToAction.StartChattingWithUser -> handleStartChatting(action)
+        }.exhaustive
+    }
+
+    private fun handleStartChatting(action: MatrixToAction.StartChattingWithUser) {
+        val mxId = action.matrixItem.id
+        val existing = session.getExistingDirectRoomWithUser(mxId)
+        if (existing != null) {
+            // navigate to this room
+            _viewEvents.post(MatrixToViewEvents.NavigateToRoom(existing))
+        } else {
+            setState {
+                copy(startChattingState = Loading())
+            }
+            // we should create the room then navigate
+            viewModelScope.launch(Dispatchers.IO) {
+                val adminE2EByDefault = rawService.getElementWellknown(session.myUserId)
+                        ?.isE2EByDefault()
+                        ?: true
+
+                val roomParams = CreateRoomParams()
+                        .apply {
+                            invitedUserIds.add(mxId)
+                            setDirectMessage()
+                            enableEncryptionIfInvitedUsersSupportIt = adminE2EByDefault
+                        }
+
+                val roomId =
+                        try {
+                            awaitCallback<String> { session.createRoom(roomParams, it) }.also {
```

Comment:
no need to use `also` here, if you move the setState() call after the `try/catch` block